back/src/Model/Space.ts

```diff
@@ -0,0 +1,139 @@
+    }
+    public updateUser(watcher: SpacesWatcher, spaceUser: PartialSpaceUser) {
+        const usersList = this.usersList(watcher);
+        const user = usersList.get(spaceUser.getUuid());
+        if (user) {
+            if (spaceUser.getTagsList().length > 0) {
+                user.setTagsList(spaceUser.getTagsList());
+            }
+            if (spaceUser.hasName()) {
+                user.setName(spaceUser.getName()?.getValue() as string);
+            }
+            if (spaceUser.hasPlayuri()) {
+                user.setPlayuri(spaceUser.getPlayuri()?.getValue() as string);
+            }
+            if (spaceUser.hasColor()) {
+                user.setColor(spaceUser.getColor()?.getValue() as string);
+            }
+            if (spaceUser.getCharacterlayersList().length > 0) {
+                user.setCharacterlayersList(spaceUser.getCharacterlayersList());
+            }
+            if (spaceUser.hasIslogged()) {
+                user.setIslogged(spaceUser.getIslogged()?.getValue() as boolean);
+            }
+            if (spaceUser.hasAvailabilitystatus()) {
+                user.setAvailabilitystatus(spaceUser.getAvailabilitystatus()?.getValue() as number);
+            }
+            if (spaceUser.hasRoomname()) {
+                user.setRoomname(spaceUser.getRoomname());
+            }
+            if (spaceUser.hasVisitcardurl()) {
+                user.setVisitcardurl(spaceUser.getVisitcardurl());
+            }
+            if (spaceUser.hasScreensharing()) {
+                user.setScreensharing(spaceUser.getScreensharing()?.getValue() as boolean);
+            }
+            if (spaceUser.hasAudiosharing()) {
+                user.setAudiosharing(spaceUser.getAudiosharing()?.getValue() as boolean);
+            }
+            if (spaceUser.hasVideosharing()) {
+                user.setVideosharing(spaceUser.getVideosharing()?.getValue() as boolean);
+            }
+            usersList.set(spaceUser.getUuid(), user);
+            const message = new UpdateSpaceUserMessage();
+            message.setSpacename(this.name);
+            message.setUser(spaceUser);
+            this.notifyWatchers(watcher, message);
+            debug(`${this.name} : user => updated ${spaceUser.getUuid()}`);
+        }
+    }
+    public removeUser(watcher: SpacesWatcher, uuid: string) {
+        const usersList = this.usersList(watcher);
+        usersList.delete(uuid);
+
+        const message = new RemoveSpaceUserMessage();
+        message.setSpacename(this.name);
+        message.setUseruuid(uuid);
+        this.notifyWatchers(watcher, message);
+        debug(`${this.name} : user => removed ${uuid}`);
+    }
+
+    public addWatcher(watcher: SpacesWatcher) {
+        this.users.set(watcher, new Map<string, SpaceUser>());
+        debug(`Space ${this.name} => watcher added ${watcher.uuid}`);
+        [...this.users.values()].forEach((spaceUsers) => {
+            [...spaceUsers.values()].forEach((spaceUser) => {
```

Comment:
I fear the `[...this.users.values()]` might be computationally expensive. You are basically creating a complete object from scratch, just to call the `forEach` method on it.

In addition, calling the `forEach` method (instead of using a `for ... of ...` loop is slower (because Javascript has to call an anonymous function on each iteration). See: https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/

Instead, use :

```
for (const spaceUsers of this.users.values()) {
  //...
}
```

Try to do this all over this PR.